cni/network/invoker_cns.go

```diff
@@ -379,32 +378,49 @@ func getRoutes(cnsRoutes []cns.Route, skipDefaultRoutes bool) ([]network.RouteIn
 	return routes, nil
 }
 
-func configureDefaultAddResult(info *IPResultInfo, addConfig *IPAMAddConfig, addResult *IPAMAddResult, overlayMode bool, key string) error {
+func configureDefaultAddResult(info *IPResultInfo, addConfig *IPAMAddConfig, addResult *IPAMAddResult, key string) error {
+	ipamMode := util.IpamMode(addConfig.nwCfg.IPAM.Mode)
 	// set the NC Primary IP in options
 	// SNATIPKey is not set for ipv6
-	if net.ParseIP(info.ncPrimaryIP).To4() != nil {
+	if ipamMode != util.Nodesubnet && net.ParseIP(info.ncPrimaryIP).To4() != nil {
 		addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
 	}
 
+	if ipamMode == util.Nodesubnet {
+		logger.Info("IPAM mode is nodesubnet, using host IP configuration info")
+		info.ncGatewayIPAddress = info.hostGateway
+		info.ncPrimaryIP = info.hostPrimaryIP
+		_, ncIPNet, err := net.ParseCIDR(info.hostSubnet)
+		if err != nil {
+			return errors.Wrap(err, "Unable to parse hostSubnet from response: "+info.hostSubnet+" with err %w")
+		}
+
+		maskLen, _ := ncIPNet.Mask.Size()
+		info.ncSubnetPrefix = uint8(maskLen) //nolint:gosec // length of mask, at most 128
+	}
+
 	ip, ncIPNet, err := net.ParseCIDR(info.podIPAddress + "/" + fmt.Sprint(info.ncSubnetPrefix))
+
 	if ip == nil || err != nil {
 		return errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
 	}
 
-	ncgw := net.ParseIP(info.ncGatewayIPAddress)
-	if ncgw == nil {
-		// TODO: Remove v4overlay and dualstackoverlay options, after 'overlay' rolls out in AKS-RP
-		if !overlayMode {
+	podGateway := net.ParseIP(info.ncGatewayIPAddress)
+	// TODO: Remove v4overlay and dualstackoverlay options, after 'overlay' rolls out in AKS-RP
+	isPodSubnetMode := (ipamMode != util.V4Overlay) && (ipamMode != util.DualStackOverlay) && (ipamMode != util.Overlay) && (ipamMode != util.Nodesubnet)
```

Comment:
could a function like: 

`func isPodSubnetMode(ipamMode string) bool {
    switch ipamMode {
    case util.V4Overlay, util.DualStackOverlay, util.Overlay, util.Nodesubnet:
        return false
    default:
        return true
    }
}`

Make this easier to read/test?

Reply:
It would, but I'm planning to finish up that TODO and get rid of two of these. Subsequently, I'm thinking of making podsubnet a separate config option, at which point this check would just be `ipamMode == util.Podsubnet`

Reply:
Right on. The (....) && (...) && (...) was a little odd, I didn't understand the need for the ()

Reply:
I'm a little paranoid that way about operator precedence :D

Reply:
`ipamMode != util.V4Overlay && 
ipamMode != util.DualStackOverlay && 
ipamMode != util.Overlay && 
ipamMode != util.Nodesubnet
`
or

`(ipamMode != util.V4Overlay && 
ipamMode != util.DualStackOverlay && 
ipamMode != util.Overlay && 
ipamMode != util.Nodesubnet)
`

are functionality equivalent :-) I will admit the current does stand out a little

Reply:
removed the parentheses